### PR TITLE
Model no memory fragment handling as wrapper PartitionMemoryEstimator

### PR DESCRIPTION
**Pull request by losipiuk**
## Description

Also add some test coverage

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/ExponentialGrowthPartitionMemoryEstimator.java [hunk 1]
```diff
@@ -117,7 +119,10 @@ void refreshNodePoolMemoryInfos()
         }
 
         @Override
-        public PartitionMemoryEstimator createPartitionMemoryEstimator(Session session, PlanFragment planFragment)
+        public PartitionMemoryEstimator createPartitionMemoryEstimator(
+                Session session,
+                PlanFragment planFragment,
+                Function<PlanFragmentId, PlanFragment> sourceFragmentLookup)
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/ExponentialGrowthPartitionMemoryEstimator.java:125]: @findepi with this lookup function we loose the caching of "no-memory" flag on child stages we used to have - should not be a big deal - as plans are not that huge and even assesing same plan fragment multiple times will be fine.

But maybe we can do better?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java [hunk 2]
```diff
@@ -1198,6 +1198,14 @@ private void createStageExecution(SubPlan subPlan, boolean rootFragment, Map<Sta
                 if (eager) {
                     sourceExchanges.values().forEach(sourceExchange -> sourceExchange.setSourceHandlesDeliveryMode(EAGER));
                 }
+
+                Function<PlanFragmentId, PlanFragment> planFragmentLookup = planFragmentId -> {
+                    StageExecution stageExecution = stageExecutions.get(getStageId(planFragmentId));
+                    if (stageExecution == null) {
+                        throw new IllegalArgumentException("stage for fragment " + planFragmentId + " not started yet");
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java:1199]: use verify

nit: use fmt

File: core/trino-main/src/test/java/io/trino/execution/scheduler/TestExponentialGrowthPartitionMemoryEstimator.java [hunk 3]
```diff
@@ -50,6 +51,10 @@
 
 public class TestExponentialGrowthPartitionMemoryEstimator
 {
+    public static final Function<PlanFragmentId, PlanFragment> THROWING_PLAN_FRAGMENT_LOOKUP = planFragmentId -> {
```
Comment (findepi) [inline core/trino-main/src/test/java/io/trino/execution/scheduler/TestExponentialGrowthPartitionMemoryEstimator.java:54]: private

File: core/trino-main/src/main/java/io/trino/execution/scheduler/ExponentialGrowthPartitionMemoryEstimator.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/ExponentialGrowthPartitionMemoryEstimator.java:125]: discussed offline. let's capture this as a TODO comment for now, as this is likely not a big deal for now.
some more comprehensive profiling of the scheduler might be useful

**Review by findepi — APPROVED** [review]

**Review by linzebing — APPROVED** [review]

**Pull request merged**